### PR TITLE
Add Spring Cloud Function build setup section

### PR DESCRIPTION
**Pull request by OlgaMaciaszek**
Fixes gh-30

### CODE REVIEW

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * @author Olga Maciaszek-Sharma
+ */
+class SpringCloudFunctionBuildSetupSection extends PreDefinedSection {
+
+	private static final String templateName = "spring-cloud-function-build-setup";
+
+	private static final String AWS_MAVEN_SETUP_EXAMPLE = "https://github.com/spring-cloud/spring-cloud-function/blob/master/spring-cloud-function-samples/function-sample-aws/pom.xml";
+
+	private static final String AWS_GRADLE_SETUP_EXAMPLE = "https://github.com/spring-cloud/spring-cloud-function/blob/master/spring-cloud-function-samples/function-sample-aws/build.gradle";
```
Comment (snicoll) [inline start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java]: I think it would be more logical if we had at least a sample per build system. Do we have a plan to improve the coverage so that at least we have a link regardless of the build system?

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionHelpDocumentCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+					new SpringCloudFunctionBuildSetupSection(CloudPlatform.AZURE,
+							this.buildSystem.id().toUpperCase(), this.templateRenderer));
+		}
+	}
+
+	/**
+	 * Contains Cloud Platform types for which additional build setup information can be
+	 * added.
+	 */
+	enum CloudPlatform {
```
Comment (snicoll) [inline start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionHelpDocumentCustomizer.java]: Do we need a type to represent this? Looks like a lot of code to me.

File: start-site/src/main/resources/templates/spring-cloud-netflix-maintenance-mode.mustache [hunk 3]
```diff
@@ -11,4 +11,4 @@ support for.
 
 Please see [this blog entry](https://spring.io/blog/2018/12/12/spring-cloud-greenwich-rc1-available-now#spring-cloud-netflix-projects-entering-maintenance-mode)
 for more information on maintenance mode and a list of suggested replacements for those
-libraries.
\ No newline at end of file
+libraries.
```
Comment (snicoll) [inline start-site/src/main/resources/templates/spring-cloud-netflix-maintenance-mode.mustache]: This is an unrelated change to the purpose of this PR. Can you please revert that? Not adding a new line is on purpose as we prefer spaces to be handled consistently when all sections are rendered.

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+	private static final String AZURE_MAVEN_SETUP_EXAMPLE = "https://github.com/spring-cloud/spring-cloud-function/blob/master/spring-cloud-function-samples/function-sample-azure/pom.xml";
+
+	private static final String FUNCTION_BUILD_SETUP_INFO = "Additional build setup for Spring Cloud Function apps - ";
+
+	SpringCloudFunctionBuildSetupSection(
+			SpringCloudFunctionHelpDocumentCustomizer.CloudPlatform platform,
+			String buildTool, MustacheTemplateRenderer templateRenderer) {
+		super(FUNCTION_BUILD_SETUP_INFO + platform.getName());
```
Comment (snicoll) [inline start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java]: I am not a huge fan of the `-` in the current title. I'd rather have something dedicated

> Build setup for Spring Cloud Function apps on Microsoft Azure
> Build setup for Spring Cloud Function apps on Amazon Web Services

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ */
+class SpringCloudFunctionBuildSetupSection extends PreDefinedSection {
+
+	private static final String templateName = "spring-cloud-function-build-setup";
+
+	private static final String AWS_MAVEN_SETUP_EXAMPLE = "https://github.com/spring-cloud/spring-cloud-function/blob/master/spring-cloud-function-samples/function-sample-aws/pom.xml";
+
+	private static final String AWS_GRADLE_SETUP_EXAMPLE = "https://github.com/spring-cloud/spring-cloud-function/blob/master/spring-cloud-function-samples/function-sample-aws/build.gradle";
+
+	private static final String AZURE_MAVEN_SETUP_EXAMPLE = "https://github.com/spring-cloud/spring-cloud-function/blob/master/spring-cloud-function-samples/function-sample-azure/pom.xml";
```
Comment (snicoll) [inline start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java]: This link is only valid for the `master` variant and would be invalid/inconsistent when a different version of Spring Cloud is selected.

**Comment by wilkinsona** [comment]
@OlgaMaciaszek Could you please rebase and force-push your branch to clean up the commits from others that aren't related to gh-30?

**Comment by OlgaMaciaszek** [comment]
@wilkinsona yes, I will; I'm still adding more changes

**Comment by OlgaMaciaszek** [comment]
Sorry, have forgotten that I'd previously already submitted this branch as PR. Still making changes. Please do not merge yet. Will comment here when ready.

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionHelpDocumentCustomizer.java (hunk 2 above)
Comment (OlgaMaciaszek) [reply start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionHelpDocumentCustomizer.java]: It makes it possible to later operate on these objects easily in `SpringCloudFunctionHelpDocumentCustomizer` and have use methods that take appropriate values from these enums rather than operating on Strings. Have a look at `addBuildSetupInfo` for example. This way, it's also extensible - we can add more cloud platforms easily, without changing the `SpringCloudFunctionHelpDocumentCustomizer` code. So, I would like to have some kind of an object here. Alternatively, we could make property-based objects and passed these values via properties. WDYT? I have opted for the enum, as I think these values will be rarely changed, though.

Comment (snicoll) [reply start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionHelpDocumentCustomizer.java]: Thanks for the feedback. Let's keep the enum and we'll figure things out when merging.

File: start-site/src/main/resources/templates/spring-cloud-netflix-maintenance-mode.mustache (hunk 3 above)
Comment (OlgaMaciaszek) [reply start-site/src/main/resources/templates/spring-cloud-netflix-maintenance-mode.mustache]: Sure. I think I was getting a checkstyle or formatting validation issue here, but not anymore.

**Comment by OlgaMaciaszek** [comment]
Hello @wilkinsona @mbhave @snicoll This is ready for review now - SC function docs for `3.0.0.M1` have been updated to contain the build setup section that is now available online, so it can be merged as well after the review passes.

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java (hunk 4 above)
Comment (OlgaMaciaszek) [reply start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java]: Not sure I understand what you mean by "dedicated"?

Comment (OlgaMaciaszek) [reply start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java]: Ok, never mind. I see that comment was for the previous version :).

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java (hunk 5 above)
Comment (OlgaMaciaszek) [reply start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java]: Outdated - changed now.

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java (hunk 1 above)
Comment (OlgaMaciaszek) [reply start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionBuildSetupSection.java]: Outdated - changed now.

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionHelpDocumentCustomizer.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		String springCloudFunctionVersion = this.projectVersionResolver.resolveVersion(bootVersion,
+				SPRING_CLOUD_FUNCTION_ARTIFACT_ID);
+		if (springCloudFunctionVersion == null) {
+			LOG.warn("Spring Cloud Function version could not be resolved for Spring Boot version: "
+					+ bootVersion.toString());
+			return;
+		}
+		if (isSnapshot(springCloudFunctionVersion)) {
+			LOG.debug("Spring Cloud Function version " + springCloudFunctionVersion
+					+ " is a snapshot. No documents are present for this version to link to.");
```
Comment (mbhave) [inline start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionHelpDocumentCustomizer.java]: Not sure I followed why the document wouldn't be there for snapshots.

File: start-site/src/main/resources/templates/spring-cloud-function-build-setup-azure.mustache [hunk 7]
```diff
@@ -0,0 +1,11 @@
+In order to run Spring Cloud Function applications on Microsoft Azure, you can leverage Spring Cloud
+Function Microsoft Azure adapter and {{buildTool}} plugins offered by the cloud platform
+provider.
+
+You will need to add to dependencies and configure the Azure Functions plugin, ensure the
```
Comment (mbhave) [inline start-site/src/main/resources/templates/spring-cloud-function-build-setup-azure.mustache]: I don't think we should mention these details here as they're all in the reference doc and we'll automatically get the updates when the reference doc is updated. WDYT?

Comment (OlgaMaciaszek) [reply start-site/src/main/resources/templates/spring-cloud-function-build-setup-azure.mustache]: I understood from a comment in a previous review that the Initializr team wanted to see some more details here. However, can change it and provide less details here.

File: start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionHelpDocumentCustomizer.java (hunk 6 above)
Comment (OlgaMaciaszek) [reply start-site/src/main/java/io/spring/start/site/extension/springcloud/SpringCloudFunctionHelpDocumentCustomizer.java]: The docs are not being published for snapshots, i.e. this link can be accessed online: https://cloud.spring.io/spring-cloud-static/spring-cloud-function/3.0.0.M1/spring-cloud-function.html , but there is no: https://cloud.spring.io/spring-cloud-static/spring-cloud-function/3.0.0.BUILD-SNAPSHOT/spring-cloud-function.html

File: start-site/src/main/resources/templates/spring-cloud-function-build-setup-azure.mustache (hunk 7 above)
Comment (mbhave) [reply start-site/src/main/resources/templates/spring-cloud-function-build-setup-azure.mustache]: Oh, I might've missed that comment. I prefer fewer details here in favor of pointing to the details in the documentation but let's see what the rest of the team thinks.

Comment (snicoll) [reply start-site/src/main/resources/templates/spring-cloud-function-build-setup-azure.mustache]: Same here. I think we should focus writing whatever is specific given the actual project in place (like if we detect you've selected A but not B we could add a blurp about the possible combination). Everything else should redirect to other resources as we want that stuff to be available to everyone, not only start.spring.io users.

Comment (OlgaMaciaszek) [reply start-site/src/main/resources/templates/spring-cloud-function-build-setup-azure.mustache]: Ok, maybe I have misunderstood the comment. Will change it now.

Comment (OlgaMaciaszek) [reply start-site/src/main/resources/templates/spring-cloud-function-build-setup-azure.mustache]: @mbhave @snicoll  done.

**Comment by mbhave** [comment]
@OlgaMaciaszek Looks good to me now. I moved it to the [Next Steps](https://github.com/mbhave/start.spring.io/commit/54a89ce14575a76bbfa565ec8dc68cc5642b9abc) section. It doesn't feel like it's own top-level section to me especially since it's possible to have multiple sections in case you add both azure and aws dependencies. I can merge it if everyone agrees with that.

**Comment by mbhave** [comment]
Thanks again @OlgaMaciaszek. This is now merged into master.

**Comment by OlgaMaciaszek** [comment]
Thanks @mbhave
